### PR TITLE
Cleanup compiler warnings

### PR DESCRIPTION
**Pull request by vmauery**
I really don't like seeing so many compiler warnings when ipmitool builds. This is the first round of removing a few of them.

### CODE REVIEW

File: include/ipmitool/helper.h [hunk 1]
```diff
@@ -38,6 +38,12 @@
 #include <stdio.h>
 #include <string.h>
 
+#ifdef __GNUC__
+#  define __UNUSED__(x) __attribute__((__unused__)) x
```
Comment (AlexanderAmelkin) [inline include/ipmitool/helper.h]: Doesn't look like this new macro is used anywhere within this pull request.
Did you forget to replace occurrences of `__attribute__((__unused__))` with it?

File: include/ipmitool/helper.h [hunk 2]
```diff
@@ -183,11 +189,23 @@ uint8_t *array_letoh(uint8_t *buffer, size_t length);
 #endif
 
 #ifndef __minlen
-# define __minlen(a, b) ({ int x=strlen(a); int y=strlen(b); (x < y) ? x : y;})
+static inline size_t min_strlen(const char *a, const char *b)
```
Comment (AlexanderAmelkin) [inline include/ipmitool/helper.h]: Yup. Already have this done in my local branch.

File: include/ipmitool/ipmi_intf.h [hunk 3]
```diff
@@ -58,7 +58,7 @@ enum LANPLUS_SESSION_STATE {
 };
 
 
-#define IPMI_AUTHCODE_BUFFER_SIZE 20
+#define IPMI_AUTHCODE_BUFFER_SIZE 20u
```
Comment (AlexanderAmelkin) [inline include/ipmitool/ipmi_intf.h]: What exactly are you addressing with this change? Doesn't look to affect the number of warnings.

File: src/plugins/ipmi_intf.c [hunk 4]
```diff
@@ -219,7 +219,7 @@ ipmi_intf_session_set_username(struct ipmi_intf * intf, char * username)
 	if (!username)
 		return;
 
-	memcpy(intf->ssn_params.username, username, __min(strlen(username), 16));
+	memcpy(intf->ssn_params.username, username, __min(strlen(username), 16u));
```
Comment (AlexanderAmelkin) [inline src/plugins/ipmi_intf.c]: Just `16` without `u` is pretty well a `size_t` value. What's the need for `u` ? There were no warnings in ipmi_intf.c even without this change.

File: lib/helper.c [hunk 5]
```diff
@@ -866,7 +867,8 @@ ipmi_start_daemon(struct ipmi_intf *intf)
 		exit(0);
 #endif
 
-	chdir("/");
+	ret = chdir("/");
+	assert(ret == 0);
```
Comment (AlexanderAmelkin) [inline lib/helper.c]: https://github.com/ipmitool/ipmitool/wiki/Coding-Standards
> Assertions
>
> `assert()` - don't; really, why would you?

Indeed, failing on an assertion is highly non-informative. Please properly check for the error and print the diagnostic message based on `errno`.

File: lib/helper.c [hunk 6]
```diff
@@ -875,9 +877,11 @@ ipmi_start_daemon(struct ipmi_intf *intf)
...
 
 	fd = open("/dev/null", O_RDWR);
-	assert(0 == fd);
-	dup(fd);
-	dup(fd);
+	assert(STDIN_FILENO == fd);
+	ret = dup(fd);
+	assert(STDOUT_FILENO == ret);
+	ret = dup(fd);
+	assert(STDERR_FILENO == ret);
```
Comment (AlexanderAmelkin) [inline lib/helper.c]: Same here. Don't use assertions. Check return values of `dup()` and print out a errno-based message on errors.

File: lib/ipmi_sdr.c [hunk 7]
```diff
@@ -204,7 +204,9 @@ sdr_convert_sensor_reading(struct sdr_record_full_sensor *sensor, uint8_t val)
 	case 1:
 		if (val & 0x80)
 			val++;
-		/* Deliberately fall through to case 2. */
+		result = (double) (((m * (int8_t) val) +
+				    (b * pow(10, k1))) * pow(10, k2));
+		break;
```
Comment (AlexanderAmelkin) [inline lib/ipmi_sdr.c]: Why? The comment is totally valid from gcc's point of view and correctly silences the implicit fall through warning.
See https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wimplicit-fallthrough

You here add unnecessary code duplication. There were no warnings in regard to this in ipmi_sdr.c. At least with gcc 5.4.0 on my Linux Mint 18.3 (Ubuntu Xenial). Wasn't it so on your system? What compiler?

File: lib/ipmi_sdr.c [hunk 8]
```diff
@@ -284,7 +286,8 @@ sdr_convert_sensor_hysterisis(struct sdr_record_full_sensor *sensor, uint8_t val
 	case 1:
 		if (val & 0x80)
 			val++;
-		/* Deliberately fall through to case 2. */
+		result = (double) (((m * (int8_t) val) ) * pow(10, k2));
+		break;
```
Comment (AlexanderAmelkin) [inline lib/ipmi_sdr.c]: same here

File: lib/ipmi_sdr.c [hunk 9]
```diff
@@ -359,7 +362,8 @@ sdr_convert_sensor_tolerance(struct sdr_record_full_sensor *sensor, uint8_t val)
 	case 1:
 		if (val & 0x80)
 			val++;
-		/* Deliberately fall through to case 2. */
+		result = (double) (((m * ((double)((int8_t) val)/2))) * pow(10, k2));
+		break;
```
Comment (AlexanderAmelkin) [inline lib/ipmi_sdr.c]: and here

File: include/ipmitool/helper.h (hunk 1 above)
Comment (vmauery) [reply include/ipmitool/helper.h]: Oops. I was originally just going to mark the unused parameters as unused, but then I decided to just remove them. I will remove this.

File: include/ipmitool/helper.h (hunk 2 above)
Comment (vmauery) [reply include/ipmitool/helper.h]: Ok, I will remove this.

File: include/ipmitool/ipmi_intf.h (hunk 3 above)
Comment (vmauery) [reply include/ipmitool/ipmi_intf.h]: By default, any literals are type int. This makes the literal an unsigned int, which then makes any comparisons with size_t safe (or at least not complain about signed/unsigned warnings)

File: lib/helper.c (hunk 5 above)
Comment (vmauery) [reply lib/helper.c]: sure.

File: lib/helper.c (hunk 6 above)
Comment (vmauery) [reply lib/helper.c]: ok

File: lib/ipmi_sdr.c (hunk 7 above)
Comment (vmauery) [reply lib/ipmi_sdr.c]: GCC removes the duplicate code and it gets rid of a warning. I am using gcc 7.3.0-27ubuntu1~18.04 on Ubuntu 18.04.

File: src/plugins/ipmi_intf.c (hunk 4 above)
Comment (vmauery) [reply src/plugins/ipmi_intf.c]: Yes, it is not needed. It is just pedantically correct :)

File: include/ipmitool/ipmi_intf.h (hunk 3 above)
Comment (AlexanderAmelkin) [reply include/ipmitool/ipmi_intf.h]: First of all this doesn't seem to actually remove any warnings. Secondly, I can't get my `gcc` to produce a warning when comparing an unsigned variable to a positive literal, can you? Lastly, if we go this way, we need to do it uniformly everywhere for all inherently unsigned literals. 
All in all, so far this looks like a superfluous change.

File: src/plugins/ipmi_intf.c (hunk 4 above)
Comment (AlexanderAmelkin) [reply src/plugins/ipmi_intf.c]: Let's not touch code without need. That hinders future blaming.

File: lib/ipmi_sdr.c (hunk 7 above)
Comment (AlexanderAmelkin) [reply lib/ipmi_sdr.c]: The concern about duplicate code is not that it occupies space. It is that it is a source of bugs. It is not uncommon when someone changes the to-be-identical code in one place and then forgets to change it in the rest of the copies.

The existing comment indeed does not work. However, even gcc 8.1.0 is happy with a `/* fall through */` comment. So please just replace it with `/* fall through */`.

Comment (vmauery) [reply lib/ipmi_sdr.c]: I am all for code depuplication. I also hate warnings when code compiles. I am of a mind of either fix the warnings or what is the point of having the compiler warn you? Really what I want is a hammer to tell the compiler that I really know what I am doing in this case to make it shut up. It looks like /* fall through */ does just that. I was not aware of this trick.

File: include/ipmitool/ipmi.h [hunk 10]
```diff
@@ -162,7 +162,7 @@ struct ipmi_rs {
 	 * Looks like this is the length of the entire packet, including the RMCP
 	 * stuff, then modified to be the length of the extra IPMI message data
 	 */
-	int data_len;
+	size_t data_len;
```
Comment (AlexanderAmelkin) [inline include/ipmitool/ipmi.h]: This change introduces like 18 new warnings. Please address them.

File: include/ipmitool/ipmi_intf.h (hunk 3 above)
Comment (AlexanderAmelkin) [reply include/ipmitool/ipmi_intf.h]: Please remove this change unless you can show that it indeed helps in any way. As I said, I don't see any related warnings and can't even reproduce one 'in-vitro'.

File: include/ipmitool/ipmi.h (hunk 10 above)
Comment (AlexanderAmelkin) [reply include/ipmitool/ipmi.h]: I know. That's why there is #13 that is still not done yet...

**Pull request merged**